Compile failure for `purescript-0.10.4`

See https://hackage.haskell.org/package/purescript-0.10.4/reports/1

I've been able to track down the problem to `bower-json` and I've updated the metadata on Hackage so users won't run into the same compilation error

https://hackage.haskell.org/package/purescript-0.10.4/revisions/

Yes, using https://hackage.haskell.org/package/purescript-0.10.4/purescript-0.10.4.tar.gz.

With Sierra (for which we use GHC 8.0.2-rc2):
https://bot.brew.sh/job/Homebrew%20Core/13988/version=sierra/console

With Yosemite and El Capitan:
```
[101 of 141] Compiling Language.PureScript.Docs.RenderedCode.Types ( src/Language/PureScript/Docs/RenderedCode/Types.hs, dist/dist-sandbox-8ca1132/build/Language/PureScript/Docs/RenderedCode/Types.o )

src/Language/PureScript/Docs/RenderedCode/Types.hs:82:5: warning: [-Wname-shadowing]
    This binding for ‘<|>’ shadows the existing binding
      imported from ‘Data.Aeson.BetterErrors’ at src/Language/PureScript/Docs/RenderedCode/Types.hs:39:1-30
      (and originally defined in ‘Data.Aeson.BetterErrors.Internal’)
[102 of 141] Compiling Language.PureScript.Docs.RenderedCode.Render ( src/Language/PureScript/Docs/RenderedCode/Render.hs, dist/dist-sandbox-8ca1132/build/Language/PureScript/Docs/RenderedCode/Render.o )
[103 of 141] Compiling Language.PureScript.Docs.RenderedCode ( src/Language/PureScript/Docs/RenderedCode.hs, dist/dist-sandbox-8ca1132/build/Language/PureScript/Docs/RenderedCode.o )
[104 of 141] Compiling Language.PureScript.Docs.Types ( src/Language/PureScript/Docs/Types.hs, dist/dist-sandbox-8ca1132/build/Language/PureScript/Docs/Types.o )

src/Language/PureScript/Docs/Types.hs:416:48: error:
    • Couldn't match type ‘[Char]’ with ‘Text’
      Expected type: String -> Either BowerError PackageName
        Actual type: Text -> Either BowerError PackageName
    • In the first argument of ‘withString’, namely ‘parsePackageName’
      In the first argument of ‘perhaps’, namely
        ‘(withString parsePackageName)’
      In the second argument of ‘key’, namely
        ‘(perhaps (withString parsePackageName))’

src/Language/PureScript/Docs/Types.hs:535:72: error:
    • Couldn't match type ‘[Char]’ with ‘Text’
      Expected type: Text -> Text
        Actual type: Text -> String
    • In the second argument of ‘(.)’, namely ‘T.unpack’
      In the second argument of ‘(.)’, namely
        ‘parsePackageName . T.unpack’
      In the first argument of ‘eachInObjectWithKey’, namely
        ‘(mapLeft ErrorInPackageMeta . parsePackageName . T.unpack)’

src/Language/PureScript/Docs/Types.hs:560:61: error:
    • Couldn't match type ‘Text’ with ‘[Char]’
      Expected type: PackageName -> String
        Actual type: PackageName -> Text
    • In the second argument of ‘(.)’, namely ‘runPackageName’
      In the first argument of ‘assocListToJSON’, namely
        ‘(T.pack . runPackageName)’
      In the second argument of ‘(.=)’, namely
        ‘assocListToJSON
           (T.pack . runPackageName)
           (T.pack . showVersion)
           pkgResolvedDependencies’
cabal: Leaving directory '.'
cabal: Error: some packages failed to install:
```

--- a/purescript.cabal
+++ b/purescript.cabal
@@ -116,7 +116,7 @@ library
                    aeson-better-errors >= 0.8,
                    ansi-terminal >= 0.6.2 && < 0.7,
                    base-compat >=0.6.0,
-                   bower-json >= 0.8,
+                   bower-json >= 1.0.0.1 && < 1.1,
                    boxes >= 0.1.4 && < 0.2.0,
                    bytestring -any,
                    containers -any,

--- a/src/Language/PureScript/Docs/Types.hs
+++ b/src/Language/PureScript/Docs/Types.hs
@@ -413,7 +413,7 @@ asReExport =
 
 asInPackage :: Parse BowerError a -> Parse BowerError (InPackage a)
 asInPackage inner =
-  build <$> key "package" (perhaps (withString parsePackageName))
+  build <$> key "package" (perhaps (withText parsePackageName))
         <*> key "item" inner
   where
   build Nothing = Local
@@ -532,7 +532,7 @@ asBookmark =
 
 asResolvedDependencies :: Parse PackageError [(PackageName, Version)]
 asResolvedDependencies =
-  eachInObjectWithKey (mapLeft ErrorInPackageMeta . parsePackageName . T.unpack) asVersion
+  eachInObjectWithKey (mapLeft ErrorInPackageMeta . parsePackageName) asVersion
   where
   mapLeft f (Left x) = Left (f x)
   mapLeft _ (Right x) = Right x
@@ -557,7 +557,7 @@ instance A.ToJSON a => A.ToJSON (Package a) where
       , "versionTag"           .= pkgVersionTag
       , "modules"              .= pkgModules
       , "bookmarks"            .= map (fmap (first P.runModuleName)) pkgBookmarks
-      , "resolvedDependencies" .= assocListToJSON (T.pack . runPackageName)
+      , "resolvedDependencies" .= assocListToJSON runPackageName
                                                   (T.pack . showVersion)
                                                   pkgResolvedDependencies
       , "github"               .= pkgGithub

--- a/src/Language/PureScript/Publish.hs
+++ b/src/Language/PureScript/Publish.hs
@@ -209,7 +209,7 @@ getBowerRepositoryInfo = either (userError . BadRepositoryField) return . tryExt
       Just Repository{..} -> do
         unless (repositoryType == "git")
           (Left (BadRepositoryType repositoryType))
-        maybe (Left NotOnGithub) Right (extractGithub (T.pack repositoryUrl))
+        maybe (Left NotOnGithub) Right (extractGithub repositoryUrl)
 
 checkLicense :: PackageMeta -> PrepareM ()
 checkLicense pkgMeta =
@@ -217,7 +217,7 @@ checkLicense pkgMeta =
     [] ->
       userError NoLicenseSpecified
     ls ->
-      unless (any isValidSPDX ls)
+      unless (any (isValidSPDX . T.unpack) ls)
         (userError InvalidLicense)
 
 -- |
@@ -320,8 +320,7 @@ asResolvedDependencies = nubBy ((==) `on` fst) <$> go
   go =
     fmap (fromMaybe []) $
       keyMay "dependencies" $
-        (++) <$> eachInObjectWithKey (parsePackageName . T.unpack)
-                                     asDependencyStatus
+        (++) <$> eachInObjectWithKey parsePackageName asDependencyStatus
              <*> (concatMap snd <$> eachInObject asResolvedDependencies)
 
 -- | Extracts only the top level dependency names from the output of
@@ -330,7 +329,7 @@ asToplevelDependencies :: Parse BowerError [PackageName]
 asToplevelDependencies =
   fmap (map fst) $
     key "dependencies" $
-      eachInObjectWithKey (parsePackageName . T.unpack) (return ())
+      eachInObjectWithKey parsePackageName (return ())
 
 asDependencyStatus :: Parse e DependencyStatus
 asDependencyStatus = do
@@ -371,7 +370,7 @@ handleDeps deps = do
       ResolvedOther _   -> (ms, pkgName : os, is)
       ResolvedVersion v -> (ms, os, (pkgName, v) : is)
 
-  bowerDir pkgName = "bower_components/" ++ runPackageName pkgName
+  bowerDir pkgName = T.unpack $ "bower_components/" <> runPackageName pkgName
 
   -- Try to extract a version, and warn if unsuccessful.
   tryExtractVersion' :: (PackageName, Text) -> PrepareM (Maybe (PackageName, Version))
@@ -406,6 +405,6 @@ getPackageName fp = do
   let xs = splitOn [pathSeparator] fp
   ys <- stripPrefix ["bower_components"] xs
   y <- headMay ys
-  case Bower.mkPackageName y of
+  case Bower.mkPackageName (T.pack y) of
     Right name -> Just name
     Left _ -> Nothing

--- a/src/Language/PureScript/Publish/BoxesHelpers.hs
+++ b/src/Language/PureScript/Publish/BoxesHelpers.hs
@@ -6,6 +6,8 @@ module Language.PureScript.Publish.BoxesHelpers
 
 import Prelude.Compat
 
+import Data.Text (Text)
+import qualified Data.Text as T
 import System.IO (hPutStr, stderr)
 
 import qualified Text.PrettyPrint.Boxes as Boxes
@@ -36,6 +38,9 @@ spacer = Boxes.emptyBox 1 1
 
 bulletedList :: (a -> String) -> [a] -> [Boxes.Box]
 bulletedList f = map (indented . para . ("* " ++) . f)
+
+bulletedListT :: (a -> Text) -> [a] -> [Boxes.Box]
+bulletedListT f = bulletedList (T.unpack . f)
 
 printToStderr :: Boxes.Box -> IO ()
 printToStderr = hPutStr stderr . Boxes.render

--- a/src/Language/PureScript/Publish/ErrorsWarnings.hs
+++ b/src/Language/PureScript/Publish/ErrorsWarnings.hs
@@ -65,7 +65,7 @@ data UserError
 
 data RepositoryFieldError
   = RepositoryFieldMissing
-  | BadRepositoryType String
+  | BadRepositoryType Text
   | NotOnGithub
   deriving (Show)
 
@@ -213,7 +213,7 @@ displayUserError e = case e of
         , "installed:"
         ])
       ] ++
-        bulletedList runPackageName (NonEmpty.toList pkgs)
+        bulletedListT runPackageName (NonEmpty.toList pkgs)
         ++
       [ spacer
       , para (concat
@@ -263,7 +263,7 @@ displayRepositoryError err = case err of
   BadRepositoryType ty ->
     para (concat
       [ "In your bower.json file, the repository type is currently listed as "
-      , "\"" ++ ty ++ "\". Currently, only git repositories are supported. "
+      , "\"" ++ T.unpack ty ++ "\". Currently, only git repositories are supported. "
       , "Please publish your code in a git repository, and then update the "
       , "repository type in your bower.json file to \"git\"."
       ])
@@ -361,7 +361,7 @@ warnNoResolvedVersions pkgNames =
       ["The following ", packages, " did not appear to have a resolved "
       , "version:"])
     ] ++
-      bulletedList runPackageName (NonEmpty.toList pkgNames)
+      bulletedListT runPackageName (NonEmpty.toList pkgNames)
       ++
     [ spacer
     , para (concat
@@ -385,7 +385,7 @@ warnUndeclaredDependencies pkgNames =
       [ "The following Bower ", packages, " ", are, " installed, but not "
       , "declared as ", dependencies, " in your bower.json file:"
       ])
-    : bulletedList runPackageName (NonEmpty.toList pkgNames)
+    : bulletedListT runPackageName (NonEmpty.toList pkgNames)
 
 warnUnacceptableVersions :: NonEmpty (PackageName, Text) -> Box
 warnUnacceptableVersions pkgs =
@@ -403,7 +403,7 @@ warnUnacceptableVersions pkgs =
       , "not be parsed:"
       ])
     ] ++
-      bulletedList showTuple (NonEmpty.toList pkgs)
+      bulletedListT showTuple (NonEmpty.toList pkgs)
       ++
     [ spacer
     , para (concat
@@ -414,7 +414,7 @@ warnUnacceptableVersions pkgs =
       ])
     ]
   where
-  showTuple (pkgName, tag) = runPackageName pkgName ++ "#" ++ T.unpack tag
+  showTuple (pkgName, tag) = runPackageName pkgName <> "#" <> tag
 
 warnDirtyWorkingTree :: Box
 warnDirtyWorkingTree =

--- a/stack-ghc-8.0.yaml
+++ b/stack-ghc-8.0.yaml
@@ -5,3 +5,4 @@ extra-deps:
 - pipes-http-1.0.2
 - wai-websockets-3.0.0.9
 - websockets-0.9.6.2
+- bower-json-1.0.0.1

--- a/stack.yaml
+++ b/stack.yaml
@@ -1,3 +1,5 @@
 resolver: lts-6.25
 packages:
 - '.'
+extra-deps:
+- bower-json-1.0.0.1

